runtime/parser/function.go

```diff
@@ -175,6 +175,122 @@ func parseParameter(p *parser) (*ast.Parameter, error) {
 	), nil
 }
 
+func parseTypeParameterList(p *parser) (*ast.TypeParameterList, error) {
+	var typeParameters []*ast.TypeParameter
+
+	p.skipSpaceAndComments()
+
+	if !p.current.Is(lexer.TokenLess) {
+		return nil, nil
+	}
+
+	startPos := p.current.StartPos
+	// Skip the opening paren
+	p.next()
+
+	var endPos ast.Position
+
+	expectTypeParameter := true
+
+	atEnd := false
+	for !atEnd {
+		p.skipSpaceAndComments()
+		switch p.current.Type {
+		case lexer.TokenIdentifier:
+			if !expectTypeParameter {
+				p.report(&MissingCommaInParameterListError{
+					Pos: p.current.StartPos,
+				})
+			}
+			typeParameter, err := parseTypeParameter(p)
+			if err != nil {
+				return nil, err
+			}
+
+			typeParameters = append(typeParameters, typeParameter)
+			expectTypeParameter = false
+
+		case lexer.TokenComma:
+			if expectTypeParameter {
+				return nil, p.syntaxError(
+					"expected type parameter or end of type parameter list, got %s",
+					p.current.Type,
+				)
+			}
```

Comment:
could abstract away these sad paths since they only differ in the expected AST element